### ISSUE TITLE
"kubectl get events" or something else should give a sensible timeline of events

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#36304]
When a user does "kubectl get events" what they probably want is a timeline of the cluster over the last N minutes. In other words, something analogous to dumping the logs from each component into a file and sorting it by timestamp. What "kubectl get events" provides today is far from this, due to compression. It may be that we won't be able to fulfill this use case until we have Infrastore, but we really need something like this for debugging.

cc/ @piosz @fgrzadkowski

**Comment by zreigz** [comment]
I can take a look for it and try prepare some proof of concept

**Comment by zreigz** [comment]
I've prepared some proposal for your request. Please take a look for my code:
https://github.com/zreigz/kubernetes/commit/0b62ecbc6d2d3321fa9d0c91d38c9087b5597d99
Now you can use command:
```
kubectl get events --since=1h
```
If you guys like it I will make PR for it

**Comment by fgrzadkowski** [comment]
I might be missing something but I think that the problem that David raised is that we are "compacting" the output of `get events` such that we say that a given event happened `N` times. It's impossible to understand the flow of what happened after what with such compaction. What we would like to do is to exactly know the order of events. I think that your code does not address this problem

**Comment by davidopp** [comment]
Right, implementing what I suggested required not compacting events. (Or at least compacting in a way that does not lose the information about interleaving with unrelated events. For example, if event X happened 10 times in a row without any other events happening in between, then it is fine to compact those 10 X events. But if X happened 5 times and then some other event Y happened and then X happened again 5 times, you cannot compact all 10 X's together; you can only compact the two blocks of 5 separately.

**Comment by jbeda** [comment]
We should be able to do this without something like infrastore.  Requiring a big external component for recent activity in the system will make small "devops-y" clusters harder to create and administer.  An Infrastore type system should be for doing advanced searching with longer epochs.

I'll add that the event compaction is lossy.  It will compact events based on Reason.  Many times (for example Deployments talking about scaling/rollouts) there is useful information in the Message field.

**Comment by smarterclayton** [comment]
Part of the reason we compact is because events without compaction don't scale well above medium sized nodes.  So if we don't have a purpose built solution events without compaction is only useful at small scales (which is not irrelevant, but limiting).  I don't think events as a history stream requires removal of compaction to be more useful.

**Comment by jbeda** [comment]
Perhaps have different levels/severity of events with some of the more "important" events being immune form compaction?  Not sure what the criteria is there but the events that kill big nodes should be on the "compact" list.  Anything that is closer to user action should probably not be compacted.  (add it to the list...)

**Comment by piosz** [comment]
Having a nice visualization of what happened in a cluster over time based on data available in events seems like a nice feature for UI.

cc @bryk @kubernetes/dashboard-maintainers

**Comment by IanLewis** [comment]
@piosz Feel free to create an issue. Would be nice to have some support for it in the API. Currently filtering has to be done client side :-|

**Comment by ankon** [comment]
What exactly is needed here?

**Comment by schollii** [comment]
aggregation/compaction should be a view-side decision, ie there should be a switch to say whether events should be compacted or not, so I could do kubectl get events --no-compaction or kubectl describe pod/whatever --events-no-compaction; if there are too many events I can easily pipe the output of a non-compacted event list and filter, sort, etc

**Comment by FKatenbrink** [comment]
Any update on this?

**Comment by Nowaker** [comment]
Any update?

**Comment by matti** [comment]
I'm surprised that
```
kubectl get events --sort-by='.metadata.creationTimestamp'
```
hasn't been mentioned yet

**Comment by zakthan** [comment]
they should be at least logged in a log file so as to be collected with a tool like elastic

**Comment by Nowaker** [comment]
Unrot please?

**Comment by pre** [comment]
>  --sort-by='.metadata.creationTimestamp'

The grouping of Events causes sort by `creationTimestamp` list events in invalid order. The creationTimestamp is when Event was seen first, but the same event might also have occurred afterwards. 

For example, an event might have the following attributess:
* `firstTimestamp: "2020-01-28T13:13:27Z"`
* `lastTimestamp: "2020-01-28T14:18:30Z"`
* `metadata.creationTimestamp: "2020-01-28T13:13:27Z"`

As a result, the output is sorted with `creationTimestamp` will have elements in random order in the `LAST SEEN`  column:

```
% kubectl get event --sort-by=metadata.creationTimestamp
LAST SEEN [...other columns omitted..] 
8m32s
3m25s
3m9s
13m
53m
```

**Comment by pre** [comment]
> kubectl get event --sort-by=lastTimestamp

To get events sorted by their last occurence, you `lastTimestamp` works *but not always*. 

Sometimes (quite often) there's an event with `"lastTimestamp": null`. This causes the command fail as a whole: 
`sorter.go:354] Field {.lastTimestamp} in [][][]reflect.Value is an unsortable type: interface, err: unsortable interface: interface`

I could live with these `"lastTimestamp": null` being sorted at the top or the bottom if all the other events would show up - without the whole command exploding.

TL;DR

It is not yet possible to get a sensible timeline of events using `kubectl get events`  with kubectl v1.17.0.

**Comment by soltysh** [comment]
We'll try to address this as part of https://github.com/kubernetes/enhancements/issues/1440

**Comment by pawelprazak** [comment]
How is this even an open issue after 4 year's is beyond me, also `lastTimestamp` gives a stack trace now:
```
% kubectl -n toolbox get events --sort-by=lastTimestamp
F1019 11:56:39.690197   19639 sorter.go:353] Field {.lastTimestamp} in [][][]reflect.Value is an unsortable type: interface, err: unsortable type: <nil>
goroutine 1 [running]:
k8s.io/kubernetes/vendor/k8s.io/klog/v2.stacks(0xc000130001, 0xc00049c0f0, 0x99, 0xeb)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:996 +0xb9
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output(0x391ec40, 0xc000000003, 0x0, 0x0, 0xc00019e070, 0x36f15ce, 0x9, 0x161, 0x0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:945 +0x191
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printf(0x391ec40, 0x3, 0x0, 0x0, 0x2867c04, 0x31, 0xc000b61710, 0x4, 0x4)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:733 +0x17a
k8s.io/kubernetes/vendor/k8s.io/klog/v2.Fatalf(...)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1456
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get.(*TableSorter).Less(0xc00063a330, 0x1, 0x0, 0x6)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get/sorter.go:353 +0x2d5
sort.doPivot(0x2a9fa00, 0xc00063a330, 0x0, 0xd, 0xc0006bbd10, 0xc00054c1a0)
	/usr/local/Cellar/go/1.15.2/libexec/src/sort/sort.go:117 +0x589
sort.quickSort(0x2a9fa00, 0xc00063a330, 0x0, 0xd, 0x8)
	/usr/local/Cellar/go/1.15.2/libexec/src/sort/sort.go:190 +0x9a
sort.Sort(0x2a9fa00, 0xc00063a330)
	/usr/local/Cellar/go/1.15.2/libexec/src/sort/sort.go:218 +0x7a
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get.(*TableSorter).Sort(...)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get/sorter.go:359
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get.(*SortingPrinter).PrintObj(0xc0006b2480, 0x2a70e80, 0xc000430cf0, 0x2a63f60, 0xc0001ca000, 0x0, 0xc0001ca000)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get/sorter.go:56 +0x20b
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get.(*TablePrinter).PrintObj(0xc0004569b0, 0x2a70fc0, 0xc000356120, 0x2a63f60, 0xc0001ca000, 0xc0007c0a20, 0xc0003e59bd)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get/table_printer.go:41 +0x1b7
k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/printers.ResourcePrinterFunc.PrintObj(0xc0006b6280, 0x2a70fc0, 0xc000356120, 0x2a63f60, 0xc0001ca000, 0x0, 0x20)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/cli-runtime/pkg/printers/interface.go:31 +0x4e
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get.(*GetOptions).Run(0xc0004f41e0, 0x2abe9e0, 0xc0001b38f0, 0xc0004f98c0, 0xc0005a4200, 0x1, 0x4, 0x0, 0x0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get/get.go:579 +0x824
k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get.NewCmdGet.func1(0xc0004f98c0, 0xc0005a4200, 0x1, 0x4)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/cmd/get/get.go:167 +0x135
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc0004f98c0, 0xc0005a41c0, 0x4, 0x4, 0xc0004f98c0, 0xc0005a41c0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:846 +0x2c2
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000162000, 0xc000132180, 0xc000132120, 0x6)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:950 +0x375
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:887
main.main()
	_output/local/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:49 +0x21d

goroutine 19 [chan receive]:
k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).flushDaemon(0x391ec40)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1131 +0x8b
created by k8s.io/kubernetes/vendor/k8s.io/klog/v2.init.0
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:416 +0xd8

goroutine 10 [select]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x290e8a0, 0x2a64540, 0xc00030c000, 0x1, 0xc00010e0c0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:167 +0x149
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x290e8a0, 0x12a05f200, 0x0, 0x1, 0xc00010e0c0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0x290e8a0, 0x12a05f200, 0xc00010e0c0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/util/logs.InitLogs
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/kubectl/pkg/util/logs/logs.go:51 +0x96

goroutine 38 [IO wait]:
internal/poll.runtime_pollWait(0x2cac17b0, 0x72, 0x2a67840)
	/usr/local/Cellar/go/1.15.2/libexec/src/runtime/netpoll.go:220 +0x55
internal/poll.(*pollDesc).wait(0xc000853718, 0x72, 0xc000d2a000, 0x4f70, 0x4f70)
	/usr/local/Cellar/go/1.15.2/libexec/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	/usr/local/Cellar/go/1.15.2/libexec/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc000853700, 0xc000d2a000, 0x4f70, 0x4f70, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.15.2/libexec/src/internal/poll/fd_unix.go:159 +0x1b1
net.(*netFD).Read(0xc000853700, 0xc000d2a000, 0x4f70, 0x4f70, 0x203000, 0x1258a1b, 0xc000a9c160)
	/usr/local/Cellar/go/1.15.2/libexec/src/net/fd_posix.go:55 +0x4f
net.(*conn).Read(0xc000356088, 0xc000d2a000, 0x4f70, 0x4f70, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.15.2/libexec/src/net/net.go:182 +0x8e
crypto/tls.(*atLeastReader).Read(0xc000569460, 0xc000d2a000, 0x4f70, 0x4f70, 0x1a, 0x35ce, 0xc000849710)
	/usr/local/Cellar/go/1.15.2/libexec/src/crypto/tls/conn.go:779 +0x62
bytes.(*Buffer).ReadFrom(0xc000a9c280, 0x2a62f40, 0xc000569460, 0x100d5e5, 0x26202e0, 0x279d7c0)
	/usr/local/Cellar/go/1.15.2/libexec/src/bytes/buffer.go:204 +0xb1
crypto/tls.(*Conn).readFromUntil(0xc000a9c000, 0x61a4008, 0xc000356088, 0x5, 0xc000356088, 0x9)
	/usr/local/Cellar/go/1.15.2/libexec/src/crypto/tls/conn.go:801 +0xf3
crypto/tls.(*Conn).readRecordOrCCS(0xc000a9c000, 0x0, 0x0, 0xc000849d18)
	/usr/local/Cellar/go/1.15.2/libexec/src/crypto/tls/conn.go:608 +0x115
crypto/tls.(*Conn).readRecord(...)
	/usr/local/Cellar/go/1.15.2/libexec/src/crypto/tls/conn.go:576
crypto/tls.(*Conn).Read(0xc000a9c000, 0xc000d08000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.15.2/libexec/src/crypto/tls/conn.go:1252 +0x15f
bufio.(*Reader).Read(0xc000599e60, 0xc000abc1f8, 0x9, 0x9, 0xc000849d18, 0x290f800, 0x15845cb)
	/usr/local/Cellar/go/1.15.2/libexec/src/bufio/bufio.go:227 +0x222
io.ReadAtLeast(0x2a62d60, 0xc000599e60, 0xc000abc1f8, 0x9, 0x9, 0x9, 0xc000118050, 0x0, 0x2a63160)
	/usr/local/Cellar/go/1.15.2/libexec/src/io/io.go:314 +0x87
io.ReadFull(...)
	/usr/local/Cellar/go/1.15.2/libexec/src/io/io.go:333
k8s.io/kubernetes/vendor/golang.org/x/net/http2.readFrameHeader(0xc000abc1f8, 0x9, 0x9, 0x2a62d60, 0xc000599e60, 0x0, 0x0, 0xc000564d80, 0x0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/frame.go:237 +0x89
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*Framer).ReadFrame(0xc000abc1c0, 0xc000564d80, 0x0, 0x0, 0x0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/frame.go:492 +0xa5
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*clientConnReadLoop).run(0xc000849fa8, 0x0, 0x0)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/transport.go:1794 +0xd8
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*ClientConn).readLoop(0xc000a9a180)
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/transport.go:1716 +0x6f
created by k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*Transport).newClientConn
	/private/tmp/kubernetes-cli-20200916-8005-u4r6v9/_output/local/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/transport.go:695 +0x66e
```

**Comment by ankon** [comment]
@pawelprazak note that your setup is outside of the [documented "version skew policy" for `kubectl`](https://kubernetes.io/docs/setup/release/version-skew-policy/#kubectl), which quite possibly could explain such stack traces.

**Comment by pre** [comment]
> @pawelprazak note that your setup is outside of the [documented "version skew policy" for `kubectl`](https://kubernetes.io/docs/setup/release/version-skew-policy/#kubectl), which quite possibly could explain such stack traces.

The stack trace above is about `lastTimestamp` being null for some of the records, which causes the crash. 

For details, see https://github.com/kubernetes/kubernetes/issues/36304#issuecomment-581864033

**Comment by pawelprazak** [comment]
yes @ankon, my bad, but in this case it does not matter, after switching to `1.18` it's still not working as expected:
```
Pawes-MacBook-Pro:azure-acr pprazak$ kubectl --context prod-westeurope get events --sort-by lastTimestamp --all-namespaces
F1020 16:07:54.407837   28946 sorter.go:353] Field {.lastTimestamp} in [][][]reflect.Value is an unsortable type: interface, err: unsortable type: <nil>
```
also look at: https://github.com/kubernetes/kubernetes/issues/36304#issuecomment-581864033

and even if it was the case that is another usability issue IMO, `kubectl` should warn on API incompatibilities when versions are too varied.

yes @pre, exactly my point, the comment you mentioned states:
> It is not yet possible to get a sensible timeline of events using kubectl get events with kubectl v1.17.0.
and it is still true.

there is no reason to not sort because of a `null` (I'm working on AKS and the nulls happen there, not sure if it's a universal issue, can't remember seeing this on EKS but that was many versions ago), it should be handled better IMO

**Comment by soltysh** [comment]
The problems with sorting events were fixed recently, so it should not fail and complain about reflection any more. For the long-term better support for events, we're trying to implement https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/20191008-kubectl-events.md

**Comment by leilajal** [comment]
@davidopp @soltysh I see there are follow up on this issue at https://github.com/kubernetes/enhancements/issues/1440. Can we close this one?

**Comment by soltysh** [comment]
Yeah, that's fair. That enhancement is covering all the topics here and is backlinking here too.
/close
